File: src/typeid2elem/decode.py
"""Composition-aware joint decoding over per-type element probabilities.

Re-ranks joint assignments (one element per type id) by:
  score = sum_t log p_model(e_t)
        + w_pmi  * mean pairwise PMI(e_i, e_j) over the DISTINCT element set
        + w_neut * charge-neutrality feasibility bonus (fraction-weighted)

Duplicate elements across type ids are allowed (split types are legitimate).
"""

from __future__ import annotations

import itertools
from pathlib import Path

import numpy as np

N_EL = 94

# Common oxidation states per Z (compact; 0 always allowed is NOT included --
# metals in alloys are handled by the |imbalance| soft penalty instead).
_OX: dict[int, tuple[int, ...]] = {
    1: (1, -1), 2: (0,), 3: (1,), 4: (2,), 5: (3,), 6: (4, 2, -4), 7: (-3, 3, 5),
    8: (-2,), 9: (-1,), 10: (0,), 11: (1,), 12: (2,), 13: (3,), 14: (4, -4),
    15: (5, 3, -3), 16: (-2, 4, 6), 17: (-1, 5, 7), 18: (0,), 19: (1,), 20: (2,),
    21: (3,), 22: (4, 3, 2), 23: (5, 4, 3, 2), 24: (3, 6, 2), 25: (2, 3, 4, 7),
    26: (2, 3), 27: (2, 3), 28: (2,), 29: (2, 1), 30: (2,), 31: (3,), 32: (4, 2),
    33: (3, 5, -3), 34: (-2, 4, 6), 35: (-1, 5), 36: (0,), 37: (1,), 38: (2,),
    39: (3,), 40: (4,), 41: (5, 3), 42: (4, 6, 3), 43: (4, 7), 44: (3, 4),
    45: (3,), 46: (2, 4), 47: (1,), 48: (2,), 49: (3, 1), 50: (2, 4),
    51: (3, 5), 52: (-2, 4, 6), 53: (-1, 5, 7), 54: (0,), 55: (1,), 56: (2,),
    57: (3,), 58: (3, 4), 59: (3,), 60: (3,), 61: (3,), 62: (3, 2), 63: (2, 3),
    64: (3,), 65: (3,), 66: (3,), 67: (3,), 68: (3,), 69: (3,), 70: (3, 2),
    71: (3,), 72: (4,), 73: (5,), 74: (4, 6), 75: (4, 7), 76: (4, 3),
    77: (3, 4), 78: (2, 4), 79: (1, 3), 80: (2, 1), 81: (1, 3), 82: (2, 4),
    83: (3, 5), 84: (-2, 2, 4), 85: (-1,), 86: (0,), 87: (1,), 88: (2,),
    89: (3,), 90: (4,), 91: (5,), 92: (4, 6), 93: (5, 4), 94: (4, 3),
}


class CompositionPrior:
    def __init__(self, costats_path: str | Path | None = None,
                 w_pmi: float = 0.5, w_neut: float = 1.5, alpha: float = 2.0):
        if costats_path is None:
            from .assets import costats_npz
            costats_path = costats_npz()
        z = np.load(costats_path)
        n = float(z["n_sets"])
        p1 = (z["single"] + alpha) / (n + alpha * N_EL)
        p2 = (z["pair"] + alpha) / (n + alpha * N_EL**2)
        self.pmi = np.log(p2) - np.log(p1[:, None]) - np.log(p1[None, :])
        self.w_pmi = w_pmi
        self.w_neut = w_neut

    def _neutrality(self, zs: tuple[int, ...], frac: np.ndarray) -> float:
        """Min |sum_t x_t * q_t| over allowed oxidation states; 0 = neutral possible."""
        best = np.inf
        for qs in itertools.product(*[_OX.get(z, (0,)) for z in zs]):
            best = min(best, abs(float(np.dot(frac, qs))))
            if best < 1e-9:
                return 0.0
        return best

    def score_extra(self, zs: tuple[int, ...], frac: np.ndarray) -> float:
        els = sorted(set(zs))
        if len(els) > 1:
            pairs = [self.pmi[a - 1, b - 1]
                     for i, a in enumerate(els) for b in els[i + 1:]]
            pmi = float(np.mean(pairs))
        else:
            pmi = 0.0
        return self.w_pmi * pmi - self.w_neut * self._neutrality(zs, frac)

    def rerank(self, logp: np.ndarray, frac: np.ndarray, top_k: int = 10,
               beam: int = 300) -> list[tuple[float, tuple[int, ...]]]:
        """logp: (T, N_EL) per-type log-probs. Returns scored assignments (desc).

        Candidates per type = its top_k elements; beam search over types keeps
        `beam` partial assignments ranked by model log-prob, then the composition
        terms re-rank the complete assignments.
        """
        T = logp.shape[0]
        cand = np.argsort(logp, axis=1)[:, ::-1][:, :top_k] + 1   # Z candidates
        partials: list[tuple[float, tuple[int, ...]]] = [(0.0, ())]
        for t in range(T):
            nxt = [(s + float(logp[t, z - 1]), zs + (int(z),))
                   for s, zs in partials for z in cand[t]]
            nxt.sort(key=lambda x: -x[0])
            partials = nxt[:beam]
        scored = [(s + self.score_extra(zs, frac), zs) for s, zs in partials]
        scored.sort(key=lambda x: -x[0])
        return scored

    def marginals(self, logp: np.ndarray, frac: np.ndarray,
                  top_k: int = 10, beam: int = 300) -> np.ndarray:
        """Per-type marginal probs from softmax over re-ranked joint assignments."""
        scored = self.rerank(logp, frac, top_k, beam)
        scores = np.array([s for s, _ in scored])
        w = np.exp(scores - scores.max())
        w /= w.sum()
        T = logp.shape[0]
        out = np.full((T, N_EL), 1e-12)
        for wi, (_, zs) in zip(w, scored):
            for t, z in enumerate(zs):
                out[t, z - 1] += wi
        return out / out.sum(1, keepdims=True)

File: metal/Sources/InaccurateRDFCalculator/Decode.swift
import Foundation

// Composition-aware joint decoding + RAPS conformal sets — a faithful port of
// src/typeid2elem/decode.py (CompositionPrior) and the conformal block of
// scripts/predict_lite.py. All math is done in Double to match numpy.
//
// The PMI matrix and calibration scalars are baked into decode_assets.json by
// metal/tools/gen_decode_assets.py (from weights/costats.npz + weights/calib.npz).

public struct CalibParams {
    public let temperature: Double
    public let qhat: Double
    public let lam: Double
    public let kReg: Int
    public let coverage: Double
}

/// Loads the bundled decode/conformal assets (PMI matrix + calibration scalars).
/// `_loaded` is a lazily-initialized static — parsed once on first use.
public enum DecodeAssets {
    public static let nEl = 94

    static func load() -> (pmi: [Double], calib: CalibParams) {
        guard let url = Bundle.module.url(forResource: "decode_assets",
                                          withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let raw = try? JSONSerialization.jsonObject(with: data),
              let obj = raw as? [String: Any],
              let pmiAny = obj["pmi"] as? [Any],
              let cal = obj["calib"] as? [String: Any] else {
            return ([Double](repeating: 0, count: 94 * 94),
                    CalibParams(temperature: 1, qhat: 1, lam: 0, kReg: 0, coverage: 0.9))
        }
        let pmi = pmiAny.map { ($0 as? NSNumber)?.doubleValue ?? 0 }
        func d(_ k: String) -> Double { ((cal[k] as? NSNumber)?.doubleValue ?? 0) }
        let calib = CalibParams(temperature: d("temperature"), qhat: d("qhat"),
                                lam: d("lam"), kReg: Int(d("k_reg")), coverage: d("coverage"))
        return (pmi, calib)
    }

    public static let _loaded = load()
    public static var calib: CalibParams { _loaded.calib }
}

/// Composition prior: re-ranks joint element assignments (one element per type
/// id) by model log-prob + PMI(co-occurrence) + charge-neutrality feasibility.
public final class CompositionPrior {
    public let pmi: [Double]                // (94*94)
    let wPmi: Double
    let wNeut: Double
    let nEl = 94

    public init(wPmi: Double = 0.5, wNeut: Double = 1.5) {
        self.pmi = DecodeAssets._loaded.pmi
        self.wPmi = wPmi
        self.wNeut = wNeut
    }

    /// Min |Σ_t x_t q_t| over allowed oxidation states; 0 => neutral achievable.
    private func neutrality(_ zs: [Int], _ frac: [Double]) -> Double {
        let options = zs.map { OXIDATION[$0] ?? [0] }
        var idx = [Int](repeating: 0, count: zs.count)
        var best = Double.greatestFiniteMagnitude
        while true {
            var dot = 0.0
            for t in 0..<zs.count { dot += frac[t] * Double(options[t][idx[t]]) }
            best = min(best, abs(dot))
            if best < 1e-9 { return 0.0 }
            // odometer increment
            var pos = zs.count - 1
            while pos >= 0 {
                idx[pos] += 1
                if idx[pos] < options[pos].count { break }
                idx[pos] = 0; pos -= 1
            }
            if pos < 0 { break }
        }
        return best
    }

    private func scoreExtra(_ zs: [Int], _ frac: [Double]) -> Double {
        let els = Array(Set(zs)).sorted()
        var pmiVal = 0.0
        if els.count > 1 {
            var acc = 0.0; var cnt = 0
            for i in 0..<els.count {
                for j in (i + 1)..<els.count {
                    acc += pmi[(els[i] - 1) * nEl + (els[j] - 1)]; cnt += 1
                }
            }
            pmiVal = acc / Double(cnt)
        }
        return wPmi * pmiVal - wNeut * neutrality(zs, frac)
    }

    /// logp: (T, 94) per-type log-probs. Returns scored assignments (desc).
    func rerank(_ logp: [[Double]], _ frac: [Double],
                topK: Int = 10, beam: Int = 300) -> [(Double, [Int])] {
        let T = logp.count
        // per-type Z candidates: top-K by log-prob
        var cand = [[Int]]()
        for t in 0..<T {
            let order = (0..<nEl).sorted { logp[t][$0] > logp[t][$1] }
            cand.append(order.prefix(topK).map { $0 + 1 })
        }
        var partials: [(Double, [Int])] = [(0.0, [])]
        for t in 0..<T {
            var nxt = [(Double, [Int])]()
            nxt.reserveCapacity(partials.count * cand[t].count)
            for (s, zs) in partials {
                for z in cand[t] { nxt.append((s + logp[t][z - 1], zs + [z])) }
            }
            nxt.sort { $0.0 > $1.0 }
            if nxt.count > beam { nxt.removeLast(nxt.count - beam) }
            partials = nxt
        }
        var scored = partials.map { (s, zs) in (s + scoreExtra(zs, frac), zs) }
        scored.sort { $0.0 > $1.0 }
        return scored
    }

    /// Per-type marginal probs (T, 94) from softmax over re-ranked assignments.
    public func marginals(_ logp: [[Double]], _ frac: [Double],
                          topK: Int = 10, beam: Int = 300) -> [[Double]] {
        let T = logp.count
        let scored = rerank(logp, frac, topK: topK, beam: beam)
        let maxS = scored.map { $0.0 }.max() ?? 0
        var w = scored.map { exp($0.0 - maxS) }
        let wsum = w.reduce(0, +)
        for i in 0..<w.count { w[i] /= wsum }
        var out = [[Double]](repeating: [Double](repeating: 1e-12, count: nEl), count: T)
        for (wi, (_, zs)) in zip(w, scored) {
            for (t, z) in zs.enumerated() { out[t][z - 1] += wi }
        }
        for t in 0..<T {
            let rowSum = out[t].reduce(0, +)
            for c in 0..<nEl { out[t][c] /= rowSum }
        }
        return out
    }
}

/// RAPS conformal prediction set + temperature scaling (predict_lite.raps_set).
public enum Conformal {
    /// Returns the conformal set (0-based class ids = Z-1) for one type's decoded
    /// probability vector, applying temperature scaling first.
    public static func set(_ probs: [Double], _ calib: CalibParams) -> [Int] {
        // temperature scaling on log-probs, then renormalize
        var z = probs.map { log($0 + 1e-12) / calib.temperature }
        let m = z.max() ?? 0
        for i in 0..<z.count { z[i] -= m }
        var e = z.map { exp($0) }
        let es = e.reduce(0, +)
        for i in 0..<e.count { e[i] /= es }
        // RAPS: greedily accumulate mass in descending prob order
        let order = (0..<e.count).sorted { e[$0] > e[$1] }
        var cum = 0.0
        var out = [Int]()
        for (i, cls) in order.enumerated() {
            cum += e[cls]
            out.append(cls)
            if cum + calib.lam * Double(max(0, i + 1 - calib.kReg)) >= calib.qhat { break }
        }
        return out
    }
}
